cylc/flow/run_modes/skip.py

```diff
@@ -0,0 +1,166 @@
+# THIS FILE IS PART OF THE CYLC WORKFLOW ENGINE.
+# Copyright (C) NIWA & British Crown (Met Office) & Contributors.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""Utilities supporting skip modes
+"""
+from logging import INFO
+from typing import (
+    TYPE_CHECKING, List, Set, Tuple)
+
+from cylc.flow import LOG
+from cylc.flow.exceptions import WorkflowConfigError
+from cylc.flow.platforms import get_platform
+from cylc.flow.task_outputs import (
+    TASK_OUTPUT_SUBMITTED,
+    TASK_OUTPUT_SUCCEEDED,
+    TASK_OUTPUT_FAILED,
+    TASK_OUTPUT_STARTED
+)
+from cylc.flow.task_state import RunMode
+
+if TYPE_CHECKING:
+    from cylc.flow.taskdef import TaskDef
+    from cylc.flow.task_job_mgr import TaskJobManager
+    from cylc.flow.task_proxy import TaskProxy
+    from typing_extensions import Literal
+
+
+def submit_task_job(
+    task_job_mgr: 'TaskJobManager',
+    itask: 'TaskProxy',
+    now: Tuple[float, str]
+) -> 'Literal[True]':
+    """Submit a task in skip mode.
+
+    Returns:
+        True - indicating that TaskJobManager need take no further action.
+    """
+    task_job_mgr._set_retry_timers(itask, itask.tdef.rtconfig)
+    itask.summary['started_time'] = now[0]
+    itask.waiting_on_job_prep = False
+    itask.submit_num += 1
+
+    itask.platform = get_platform()
+    itask.platform['name'] = RunMode.SKIP
+    itask.summary['job_runner_name'] = RunMode.SKIP
+    itask.tdef.run_mode = RunMode.SKIP
```

Comment:
This is changing the run mode on the `TaskDef` object (which applies to all instances of a task) not the `TaskProxy` object (which only applies to a specific instance of a task).

In theory two instances of the same task could be submitted at the same time, but with different modes (quite likely if a cycle is being skipped). Because both will be using the same `TaskDef`, these submissions could interact via this shared run mode.